### ISSUE TITLE
Assume that array lookup can return undefined inside conditional expressions

### ISSUE THREAD

**Issue by ifeltsweet** [microsoft/TypeScript#41872]
## Search Terms

Array lookup
Array lookup undefined
Array indexing undefined

## Suggestion

Please forgive my TypeScript terminology.

I am proposing for the compiler to assume that array lookup can return undefined inside conditional expressions ONLY.

I am NOT proposing for all array lookup to return `T | undefined` (#11122)

## Use Cases

This helps to catch subtle, hard to spot bugs.

## Examples

```ts
const setItem = (value: Item | undefined) => { item = value };

// This a bug, setItem shouldn't accept null, TS didn't catch it.
setItem(response.data.items[0] || null)
```

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by MartinJohns** [comment]
Your example doesn't compile:

    type Item = string
    type ResponseData = { data: { items: Item[] } }

    let item: Item | undefined
    const setItem = (value: Item | undefined) => { item = value };

    const response: ResponseData = { data: { items: [] } }

    setItem(response.data.items[0] || null)

[Playground link](https://www.typescriptlang.org/play?ssl=10&ssc=1&pln=1&pc=1#code/C4TwDgpgBAksEFsoF4oGdgCcCWA7A5gFCiRQBKEaYA9rmhACICGwTKUA3lACYtMBcnKNngI0guIgDaAXSgBfBYUIAbCMGGiJoqAB8oAV1zcIAMzwRuhAMa0M6dZKSoAFADcmKgxG2I9h4zMLbgBKFAA+IRE-VA8vaHkAbmVbOg1MSho6H3JMu0Y+di5eVkEuaLFBWQUlQnpgJxcMqnyAOhKmVoq0KQAGOV19XAMVFRDCIA)

**Comment by ifeltsweet** [comment]
@MartinJohns, thanks for pointing this out.

`Item` must be a non-primitive type for the example to work. If `Item` is a primitive type then for some reason compiler does catch it.

```ts
type Item = { name: string };
type ResponseData = { data: { items: Item[] } };

let item: Item | undefined;
const setItem = (value: Item | undefined) => { item = value };

const response: ResponseData = { data: { items: [] } };

setItem(response.data.items[0] || null);
```

[Playground link](https://www.typescriptlang.org/play?ssl=1&ssc=29&pln=1&pc=13#code/C4TwDgpgBAksEFsoF4oG8oDsCGCIC4oBnYAJwEtMBzKAXwG4AoUSKAJQiLAHtMiIAItmDYU6KABNh2QhnLwERQnEQBtALp06TRgBsIwKPMTKFUAD5QArpgkQAZpQgSmAY14liBlUlQAKADdsXSsCWDNLGztHTGcAShQAPnFjXyggkOgGRkZ3PkNSTh4+MI4uD0FpMQwpEVkjBSUoDS1sxn5gHz9C8pKAOlrsPtSiVQAGTXNLTCtdXTimIA)

**Comment by ifeltsweet** [comment]
Turning on `noUncheckedIndexedAccess` makes the compiler raise the error. However `noUncheckedIndexedAccess` is an all-in flag. This feature request is for conditional expressions only.

Considering that primitive types do raise the error without this flag, I am wondering whether this is a bug or not.

**Comment by RyanCavanaugh** [comment]
It's very, very confusing in programming languages when expressions change behavior based on where you write them, so we don't do this unless the upside is extremely compellilng.

> Considering that primitive types do raise the error without this flag, I am wondering whether this is a bug or not.

`"" | null` is `null`, as is `0 || null` so the expression `response.data.items[0] || null` can produce `null` if the left operand is `string` or `number`

**Comment by ifeltsweet** [comment]
Thanks for the explanation about why the primitives work in this case.

You can close this issue if something like this is unlikely to be implemented. Those who want more type safety will have to opt in to `noUncheckedIndexedAccess`. Cheers.
